web/src/input_pill.ts

```diff
@@ -5,12 +5,56 @@ import $ from "jquery";
 import render_input_pill from "../templates/input_pill.hbs";
 
 import * as blueslip from "./blueslip";
+import type {EmojiRenderingDetails} from "./emoji";
 import * as keydown_util from "./keydown_util";
 import * as ui_util from "./ui_util";
 
 // See https://zulip.readthedocs.io/en/latest/subsystems/input-pills.html
 
-export function create(opts) {
+export type PillItem<T> = {
+    display_value: string;
+    type: string;
+    img_src?: string;
+    deactivated?: boolean;
+    status_emoji_info?: EmojiRenderingDetails & {emoji_alt_code: boolean}; // TODO: Move this in user_status.js
+} & T;
+
+type CreateOptions<T> = {
+    $container: JQuery;
+    pill_config?: {
+        show_user_status_emoji?: boolean;
+    };
+    create_item_from_text: (text: string, existing_items: PillItem<T>[]) => PillItem<T> | undefined;
+    get_text_from_item: (item: PillItem<T>) => string;
+};
+
+type Pill<T> = {
+    item: PillItem<T>;
+    $element: JQuery;
+};
```

Comment:
I think I'd prefer names like `InputPill`, `InputPillStore`, etc.; there's no real harm in extra words, and aligning with the name of the module seems more readable.